Change object cache algorithm for DSN v3

shamil-gadelshin opened the pull request:
This PR changes the piece cache forming algorithm:
- remove announcements from object cache
- remove provider record processing for object cache
- modify Node RPC:
  - change segment notification to segment header notification
  - add `get_piece` RPC method
- add new segment index processing for populating farmer's piece cache from the local node
- add local piece cache restoration on the farmer's restart

Fixes https://github.com/subspace/subspace/issues/1499 Fixes https://github.com/subspace/subspace/issues/1475

### Code contributor checklist:
* [x] I have read, understood and followed [contributing guide](https://github.com/subspace/subspace/blob/main/CONTRIBUTING.md)

shamil-gadelshin committed:
Change pieces` notification for nodes.

farmer: Add piece cache restoration.

farmer: Refactor solution.

nazar-pc commented on crates/sc-consensus-subspace-rpc/src/lib.rs:
```diff
@@ -103,6 +104,9 @@ pub trait SubspaceRpcApi {
     ) -> RpcResult<Vec<Option<SegmentHeader>>>;
+
+    #[method(name = "subspace_Piece")]
+    async fn piece(&self, piece_index: PieceIndex) -> RpcResult<Option<Vec<u8>>>;
```
This is more appropriate for RPC, you even called it `get_piece` in PR description :slightly_smiling_face: 
```suggestion
    #[method(name = "subspace_getPiece")]
    async fn get_piece(&self, piece_index: PieceIndex) -> RpcResult<Option<Vec<u8>>>;
```

It should also be non-async and marked as blocking due to the fact that it is most likely going to read something from the database: https://docs.rs/jsonrpsee-proc-macros/0.18.2/jsonrpsee_proc_macros/attr.rpc.html#method-attribute

shamil-gadelshin replied:
Non-async and blocking - make sense. As to renaming - we have "segment_headers" and other methods with "entity()" syntax. Please, confirm you want an exception here, I thought our RPC code follows rust conventions for methods: https://rust-lang.github.io/api-guidelines/naming.html.

nazar-pc replied:
Okay, just `fn piece` makes sense to me, but RPC methods in this file all need to be corrected:
* `subspace_recordsRoots` -> `subspace_segmentCommitments` (also likely redundant considering we have the next one)
* `subspace_SegmentHeaders` -> `subspace_segmentHeaders`
* `subspace_Piece` -> `subspace_piece`

shamil-gadelshin replied:
https://github.com/subspace/subspace/issues/1545

nazar-pc commented on crates/subspace-farmer-components/src/piece_caching.rs:
```diff
@@ -6,8 +6,8 @@ use subspace_core_primitives::{Piece, PieceIndexHash};
 // TODO: Re-think this number
-const CACHE_ITEMS_LIMIT: NonZeroUsize = NonZeroUsize::new(2000)
-    .expect("Archived history segment contains at very least one piece; qed");
+const CACHE_ITEMS_LIMIT: NonZeroUsize =
+    NonZeroUsize::new(100).expect("Archived history segment contains at very least one piece; qed");
```
Do we need this cache at all? We are not announcing pieces anymore. And for other purposes 100 pieces is too little to be useful (it is both less than a segment and less than a sector). I think we should just remove it now. Maybe in a separate PR.

shamil-gadelshin replied:
https://github.com/subspace/subspace/issues/1533

nazar-pc commented on crates/subspace-farmer/src/bin/subspace-farmer/commands/farm.rs:
```diff
@@ -144,10 +139,25 @@ where
+        let node_client = node_client.clone();
+
+        fill_piece_cache_from_archived_segments(run_cache_sync_tx, node_client, piece_cache)
+    });
```
Ideally these would be cancellable. When future returned by `farm_multi_disk` is dropped, these should stop even if they didn't finish. This is in binary code, so not critical at all here and can be left as is, I'm just worried SDK might later copy this code as is and will have issues with restarts.

cc @i1i1

nazar-pc replied:
Also these should be spawned onto a separate thread since cache writing will use disk I/O (there is `run_future_in_dedicated_thread` utility function in farmer already for this purpose).

shamil-gadelshin replied:
Ok. Changed to a dedicated thread for now.

nazar-pc commented on crates/subspace-farmer/src/bin/subspace-farmer/commands/farm.rs:
```diff
@@ -431,3 +441,141 @@ fn derive_libp2p_keypair(schnorrkel_sk: &schnorrkel::SecretKey) -> Keypair {
+    let piece_getter = piece_getter.clone();
+
+    if let Some(segment_index) = run_cache_sync_rx.recv().await {
+        run_cache_sync_rx.close();
```
Just curious, why do you need to close it explicitly and early, it'll be dropped anyway (comment would be helpful in such not very obvious cases).

shamil-gadelshin replied:
There is a check on the sender part for the closed channel. We need a single message here. I will add a comment.

nazar-pc commented on crates/subspace-farmer/src/bin/subspace-farmer/commands/farm.rs:
```diff
@@ -144,10 +139,25 @@ where
+        let node_client = node_client.clone();
+
+        fill_piece_cache_from_archived_segments(run_cache_sync_tx, node_client, piece_cache)
+    });
```
Looks like these two function with a channel between them should have been just a single function or one function spawned from within another function. Channel seems unnecessary here.

shamil-gadelshin replied:
The channel seems unnecessary indeed (we can get the latest segment index using history size from farmer's info). However, I would keep those functions separated for simplicity. It's easier to reason about separate processes.

nazar-pc commented on crates/subspace-farmer/src/bin/subspace-farmer/commands/farm.rs:
```diff
@@ -431,3 +441,141 @@ fn derive_libp2p_keypair(schnorrkel_sk: &schnorrkel::SecretKey) -> Keypair {
+/// Restores piece cache on startup. It waits for the new segment index and check all pieces from
+/// previous segments to see if they are already in the cache. If they are not, they are added
+/// from DSN.
+async fn restore_pieces_cache<PG, PC>(
```
A few thoughs:
1) This is not really a restoration, I'd call it "populate cache"
2) We don't need to wait for new segment to be archived, it is unnecessary, we already know about the history from `farmer_app_info.protocol_info.history_size`, no need to wait flor new segment to be archived

nazar-pc commented on crates/subspace-networking/src/create.rs:
```diff
@@ -270,6 +270,7 @@ where
             .set_record_filtering(KademliaStoreInserts::FilterBoth)
             // Providers' settings
             .set_provider_record_ttl(KADEMLIA_PROVIDER_TTL_IN_SECS)
+            // TODO: remove republication after removing all pieces` announcements
```
What does this mean? Is there still something left?

shamil-gadelshin replied:
I will remove it after changing the archival storage part.

nazar-pc commented on crates/subspace-farmer/src/bin/subspace-farmer/commands/farm.rs:
```diff
@@ -431,3 +441,141 @@ fn derive_libp2p_keypair(schnorrkel_sk: &schnorrkel::SecretKey) -> Keypair {
+        for piece_index in 0..segment_index.first_piece_index().into() {
+            let key = PieceIndex::from(piece_index).hash().to_multihash().into();
+
+            let should_cache = { piece_cache.lock().await.should_cache(&key) };
```
Just a though, it would have been faster to not take async lock every time, we could take a lock once and quickly skip through all first pieces that we already have, which should be most of them. We can even parallelize that (doing processing in batches of segment size).

shamil-gadelshin replied:
Not sure I follow here 100%. Did you mean - to introduce a loop and break it when we encounter the first "should cache" item?

nazar-pc replied:
Yes, that was the idea

nazar-pc commented on crates/subspace-farmer/src/bin/subspace-farmer/commands/farm.rs:
```diff
@@ -431,3 +441,141 @@ fn derive_libp2p_keypair(schnorrkel_sk: &schnorrkel::SecretKey) -> Keypair {
+                }
+
+                debug!(%segment_index, "Finished processing archived segment.");
+            }
```
This is a little problematic IMHO. This assumes node doesn't produce segments faster than farmer processes them, which may or may not (during node sync) be true. When it is not true, notification stream above can exceed its size and drop RPC connection (IIRC), also pieces here will not be guaranteed to be available anymore.

I'm thinking we should probably re-introduce archiving acknowledgement here so we have a feedback loop from farmer back to the node.

Then we should only need one attempt and it should really be a hard error that piece is not there anymore right after it was archived.

shamil-gadelshin replied:
I restored the code we discussed previously from this commit https://github.com/subspace/subspace/commit/08b6ee400e10adbd7b1cdcf4df2b7eb0091fec65#diff-38362e1fdf3d3ede40d53ae7c45b7f7cb969a580b28c2adf80470b407cdcd8e3

I don't 100% follow the previous solution since it seems to have several drawbacks, the most important appears to be node hangs after farmers restart. It looks like we need to introduce some periodic cleanup for the obsolete acknowledgments. Also, the acknowledgment doesn't guarantee piece availability, we need to make a more complex workflow that seems overkill here to avoid retries. Did I understand you correctly? I welcome your comments here.

nazar-pc replied:
Yes, it can hang on farmer disconnection if it happens in the middle of segment processing. This can be fixed though, you need to introduce a data structure that will be owned by the stream and will remove corresponding senders when dropped (it will be dropped when stream is dropped, which happens on disconnection). Not sure if jsonrpsee has a nicer API for this, didn't check.

shamil-gadelshin replied:
I found an easier way: remove the subscription state after the `pipe_from_stream` call.

nazar-pc commented on crates/subspace-farmer/src/bin/subspace-farmer/main.rs:
```diff
@@ -77,7 +77,7 @@ struct DsnArgs {
     listen_on: Vec<Multiaddr>,
     /// Piece cache size in pieces.
-    #[arg(long, default_value = "65536")]
+    #[arg(long, default_value = "1000")]
```
Fine for now, but this should rather be something like 1% of plotted space (and spread across multiple plots). Please create an issue for now if we don't have it yet, I'll elaborate on the design I imagine there.

shamil-gadelshin replied:
https://github.com/subspace/subspace/issues/1534

nazar-pc commented on crates/subspace-farmer/src/utils/farmer_piece_getter.rs:
```diff
@@ -43,41 +34,18 @@ where
             }
-
-            piece_cache.should_cache(&key)
         };
```
```suggestion
        if let Some(piece) = self.piece_cache.lock().await.get_piece(&key) {
            return Ok(Some(piece));
        }
```

nazar-pc commented on crates/subspace-service/src/lib.rs:
```diff
@@ -559,11 +555,11 @@ where
             node,
             bootstrap_nodes,
-        } => (node, bootstrap_nodes),
+        } => (node, bootstrap_nodes, None),
```
I don't think this is correct. If piece cache size is a function of the plot size, then even if networking is provided from the outside, piece cache should still work and be functional, it'll just be potentially composed of node + farmer cache sizes (but they will cache different pieces due to different policies).

Let's have at least a TODO here because this is problematic, it seems to make SDK and its users incredibly useless for caching purposes on the network.

shamil-gadelshin replied:
I expected change requests from @i1i1 when we stabilize the new implementation and get more info about SDK requirements. I'll add TODO.

nazar-pc replied:
Not sure what you mean about SDK requirements, this is a core protocol requirement: regardless of how you run a farmer, it should provide piece cache, but now it will not in case of SDK unless I misunderstand something.

shamil-gadelshin replied:
AFAIK, @i1i1 changes piece cache implementation for CLI. I would wait until we know how to set this parameter in our use case.

nazar-pc reviewed (commented):
A few comments and concerns. I like the simplifications, but I wish there were a few commits here (segment header refactoring and RPC method addition didn't have to be squashed with the rest of the changes).

shamil-gadelshin committed:
Refactor pieces cache.

shamil-gadelshin committed:
farmer: Simplify segment index retrieval.

Change segment index retrieval for piece cache population (from last archived segment to current history size).

shamil-gadelshin committed:
Restore farmer acknowledgement for archived segments.

nazar-pc commented on crates/subspace-farmer/src/bin/subspace-farmer/commands/farm.rs:
```diff
@@ -452,53 +457,46 @@ fn derive_libp2p_keypair(schnorrkel_sk: &schnorrkel::SecretKey) -> Keypair {
+    let final_piece_index =
+        u64::from(segment_index.first_piece_index()) + ArchivedHistorySegment::NUM_PIECES as u64;
+
+    for piece_index in 0..final_piece_index {
```
Hm... this means we'll start from scratch on every restart and it will take some time to go through all indices. Let's have at least a TODO since I expect this to be a usability problem.

Also I believe you meant this:
```suggestion
    for piece_index in 0..=final_piece_index {
```

shamil-gadelshin replied:
Actually, I believe the edge case is set correctly. I think the misunderstanding originated in the incorrect naming. WDUT about `final_piece_index` -> `piece_index_limit` ?

nazar-pc replied:
The naming is the issue indeed. It is not the final index, it is the index right after last index. I'd use `(0..).take(num_piece_indices)` instead to clearly show what you want to do here.

shamil-gadelshin replied:
The code changed. I keep `final_piece_index for now.

nazar-pc commented on crates/subspace-farmer/src/bin/subspace-farmer/commands/farm.rs:
```diff
@@ -142,14 +141,20 @@ where
+        .map_err(|err| anyhow::anyhow!("Couldn't receive history size: {}", err,))?
+        .protocol_info
+        .history_size
+        .segment_index();
```
`farmer_app_info` is already retrieved on line 89 above, no need to make another RPC query:
```suggestion
    let last_segment_index = farmer_app_info
        .protocol_info
        .history_size
        .segment_index();
```

nazar-pc commented on crates/sc-consensus-subspace/src/archiver.rs:
```diff
@@ -556,7 +557,10 @@ async fn send_archived_segment_notification(
 
-    while acknowledgement_receiver.next().await.is_some() {
-        // Wait until all acknowledgements are received
+    if acknowledgement_receiver.next().await.is_some() {
```
Why was this changed to `if`? There could be multiple acknowledgements and we need to wait for all of them, this is why we had `while` and corresponding comment.

shamil-gadelshin replied:
I wonder how it worked previously. Without changing to `if` the code hangs even after all the acknowledgments (I found two on the node side). Now, it works "after the first farmer acknowledgment".

nazar-pc replied:
If it hangs, this only means that something is still holding sender (but shouldn't), thus there is a bug that needs to be fixed and `if` masks this bug instead and violates documented behavior (that block import will not proceed until all receivers acknowledged receiving of the notification).

We had bugs related to this before due to incorrect usage, maybe there is incorrect usage now as well.

nazar-pc commented on crates/subspace-farmer/src/bin/subspace-farmer/commands/farm.rs:
```diff
@@ -511,7 +513,7 @@ async fn fill_piece_cache_from_archived_segments(
                 let segment_index = segment_header.segment_index();
 
-                debug!(%segment_index, "Starting to process archived segment....");
+                info!(%segment_index, "Starting to process archived segment....");
```
Helpful for debugging, not helpful for regular users.

```suggestion
                debug!(%segment_index, "Starting to process archived segment....");
```

nazar-pc commented on crates/subspace-farmer/src/bin/subspace-farmer/commands/farm.rs:
```diff
@@ -565,7 +567,19 @@ async fn fill_piece_cache_from_archived_segments(
+                        debug!(%segment_index, "Acknowledged archived segment.");
+                    }
+                    Err(err) => {
+                        error!(%segment_index, ?err, "Failed to acknowledged archived segment.");
```
```suggestion
                        error!(%segment_index, ?err, "Failed to acknowledge archived segment.");
```

nazar-pc commented on crates/subspace-farmer/src/bin/subspace-farmer/commands/farm.rs:
```diff
@@ -144,19 +144,25 @@ where
-    tokio::spawn({
-        let piece_cache = piece_cache.clone();
-        let piece_getter = piece_getter.clone();
+    tokio::spawn(run_future_in_dedicated_thread(
```
You don't need to spawn it, instead, if you don't want to wait for the finish, just store it in local variable. That way when this future (returned by `farm_multi_disk`) is dropped it will also cancel corresponding task and stop dedicated thread automatically.

Look at `run_future_in_dedicated_thread` implementation.

shamil-gadelshin committed:
farmer: Refactor piece cache population.

shamil-gadelshin committed:
Merge branch 'main' into new-piece-cache

nazar-pc reviewed (commented):
Last issue to resolve if `if` that must be `while`

shamil-gadelshin committed:
Merge branch 'main' into new-piece-cache

# Conflicts:
#	crates/subspace-service/Cargo.toml

shamil-gadelshin committed:
Change archived segment acknowledgment.

nazar-pc reviewed (approved):
> Chain lookup failed: No block weight for parent header.

This is due to changed fork choice rule and edge-cases that appear on new chain with max solution range. Use `SolutionRange::MAX - 100_000` and it should work fine.
